Add new model provider Novita AI

jasonhp opened the pull request:
## Title

Add new model provider Novita AI

## Type

🆕 New Feature
📖 Documentation

## Changes

- Added new model provider Novita AI
- Added related docs.

## [REQUIRED] Testing - Attach a screenshot of any new tests passing locally
If UI changes, send a screenshot/GIF of working UI fixes

<img width="1111" alt="image" src="https://github.com/user-attachments/assets/6ab89c85-ff9d-4443-a628-99742e3cf08f" />
<img width="1107" alt="image" src="https://github.com/user-attachments/assets/5263d398-8809-409b-9487-719f59200169" />

jasonhp committed:
feat: add new model provider Novita AI

jasonhp committed:
feat: sync latest code

jasonhp committed:
Merge branch 'main' into feat/novita

jasonhp committed:
Merge branch 'main' into feat/novita

jasonhp commented:
@ishaan-jaff  Hi! I believe it’s ready for review and would be great to have it merged soon if everything looks good to you. Please let me know if there’s anything I should adjust.

jasonhp committed:
feat: use deepseek r1 model for examples in Novita AI docs

jasonhp committed:
Merge branch 'main' into feat/novita

jasonhp committed:
Merge branch 'main' into feat/novita

krrishdholakia commented on litellm/main.py:
```diff
@@ -2973,6 +2973,49 @@ def completion(  # type: ignore # noqa: PLR0915
                 )
                 return response
             response = model_response
+        elif custom_llm_provider == "novita":
```
there is no need for this if block, just register as `litellm.openai_compatible_providers` in init

and it will route it into the openai block

jasonhp replied:
We need to add a custom header in this if block. I'm not sure where else to add it.

krrishdholakia replied:
you should add it in the validate_environment in the config, that's where headers can be set.

krrishdholakia replied:
in this case, it would be better to have it be like openrouter, and use base_llm_http_handler.py 

https://github.com/BerriAI/litellm/blob/4a6dcd64ff1a6151d7e5fec022b3905265791fcd/litellm/main.py#L2286

This is where we are moving all llm calls to - this way all your custom logic exists only within your transformation.py file

jasonhp replied:
Fixed

krrishdholakia commented on litellm/main.py:
```diff
@@ -3923,6 +3965,7 @@ async def atext_completion(
             or custom_llm_provider == "huggingface"
             or custom_llm_provider == "ollama"
             or custom_llm_provider == "vertex_ai"
+            or custom_llm_provider == "novita"
```
no need for this, see line below - just register as an openai_compatible_provider

jasonhp replied:
Fixed

krrishdholakia commented on litellm/utils.py:
```diff
@@ -2879,6 +2882,7 @@ def get_optional_params(  # noqa: PLR0915
             and custom_llm_provider != "bedrock"
             and custom_llm_provider != "ollama_chat"
             and custom_llm_provider != "openrouter"
+            and custom_llm_provider != "novita"
```
no need for this, just register as an openai compatible provider

jasonhp replied:
Fixed

krrishdholakia commented:
Hi @jasonhp thank you for your PR 

1. Please resolve merge conflicts
2. Please follow the updated contribution guidelines - https://docs.litellm.ai/docs/extras/contributing_code

I would expect to see 
1. Your e2e mock test added here - https://github.com/BerriAI/litellm/blob/fcbed43f0ca1ea1a0828d48551673c9ac1ba51ca/tests/litellm/test_main.py#L4
2. any additional unit tests added in the corresponding `test_` file,  in here https://github.com/BerriAI/litellm/tree/main/tests/litellm
3. A screenshot of your pr passing all tests from `make test-unit`

jasonhp committed:
Merge branch 'main' into feat/novita

jasonhp committed:
fix: fix tests

jasonhp committed:
fix: fix tests for novita

jasonhp committed:
fix: fix novita transformation

jasonhp committed:
Merge branch 'main' into feat/novita

jasonhp commented on litellm/llms/novita/chat/transformation.py:
```diff
@@ -0,0 +1,24 @@
+        model: str,
+        drop_params: bool,
+    ) -> dict:
+        mapped_openai_params = super().map_openai_params(
```
Fixed

jasonhp commented:
@krrishdholakia Hello, thank you for reviewing. I have fixed the problems you mentioned. And I fixed all unit tests and added e2e tests. I have run `make test-unit` successfully.

<img width="1412" alt="image" src="https://github.com/user-attachments/assets/f95b9d51-ffff-40c8-98d6-90989bafb685" />

jasonhp committed:
Merge remote-tracking branch 'upstream/litellm_contributor_prs_03_24_2025_p1' into feat/novita

jasonhp commented:
@krrishdholakia Hello! All conflicts have been resolved.

krrishdholakia commented:
Yup - thank you for your work on this

jasonhp commented:
@krrishdholakia Hello! I wonder when this `litellm_contributor_prs_03_24_2025_p1` branch can be merged into main branch and be released into production package?

viktor2077 commented:
@krrishdholakia Hi Krish, can you help merge these changes to the main branch?
